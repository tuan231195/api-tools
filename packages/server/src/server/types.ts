export type Config = {
	port?: number;
	baseUrl?: string;
	overrides?: string;
	routes?: string;
	init?: string;
	uri: string;
	seed?: string;
	validate?: boolean | ValidateOptions;
};

export type ValidateOptions = {
	validateRequests?: boolean;
	validateResponses?: boolean;
};

export const defaultValidateOptions: ValidateOptions = {
	validateRequests: true,
	validateResponses: false,
};

export class InvalidRequestError extends Error {
	constructor(public readonly errors: ValidationErrorItem[]) {
		super('Invalid Request Error');
	}
}

export class InvalidResponseError extends Error {
	constructor(public readonly errors: ValidationErrorItem[]) {
		super('Invalid Response Error');
	}
}

interface ValidationErrorItem {
	message: string;
	error_code?: string;
	path?: string;
}
