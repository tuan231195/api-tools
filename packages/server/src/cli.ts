#! /usr/bin/env node
require('ts-node/register/transpile-only');

import yargs from 'yargs';

yargs
	.scriptName('mock-swagger-server')
	.command(require('./cli/index').default)
	.strict(true)
	.showHelpOnFail(false);

if (require.main === module) {
	yargs.argv;
}

export default yargs;
