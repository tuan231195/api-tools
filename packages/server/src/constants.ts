export const NAMESPACE = 'api-tools:mock-server';
