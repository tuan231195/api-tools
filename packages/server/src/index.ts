require('ts-node/register/transpile-only');
export * from './server';
