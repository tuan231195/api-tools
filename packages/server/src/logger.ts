import debug from 'debug';

import { NAMESPACE } from 'src/constants';

const logger = debug(NAMESPACE);

export default logger;
