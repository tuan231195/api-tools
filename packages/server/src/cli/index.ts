import { CommandModule } from 'yargs';
import path from 'path';
import { startServer } from 'src/server';
import { initSeed } from '@vdtn359/api-tools-faker';

const commandModule: CommandModule<
	{},
	{
		configFile: string;
		seed?: string;
	}
> = {
	describe: 'run mock swagger server',
	handler: async function ({ configFile, seed }) {
		if (!path.isAbsolute(configFile)) {
			configFile = path.resolve(process.cwd(), configFile);
		}
		return startServer(configFile, seed);
	},
	command: '$0',
	builder: {
		configFile: {
			demandOption: true,
			type: 'string',
			description: 'The config file',
			alias: 'c',
		},
		seed: {
			demandOption: false,
			type: 'string',
			description: 'The seed. Used for predictable response',
			alias: 's',
		},
	},
};

export default commandModule;
