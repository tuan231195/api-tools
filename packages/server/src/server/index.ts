import { setupServer } from 'src/server/server';
import path from 'path';
import { Config } from 'src/server/types';
import fs from 'fs';

export const initServer = async (configFile: string, seed?: string) => {
	const config = parseConfig(configFile);

	return setupServer({
		...config,
		seed,
	});
};

export const startServer = async (configFile: string, seed?: string) => {
	const app = await initServer(configFile, seed);

	app.listen(app.get('port'), function () {
		console.info(`Express server listening on port ${app.get('port')}`);
	});
};

const parseConfig = (configFile: string): Config => {
	const config = require(configFile);
	const dirname = path.dirname(configFile);

	if (!config.uri) {
		throw new Error('uri is required');
	}

	return {
		...config,
		init: resolveFile(dirname, config.init),
		uri: resolveFile(dirname, config.uri, true),
		baseUrl: config.baseUrl ?? '/',
		overrides: resolveFile(dirname, config.overrides),
		routes: resolveFile(dirname, config.routes),
	};
};

const resolveFile = (dirname: string, file?: string, allowUri = false) => {
	if (!file) {
		return file;
	}

	if (allowUri && isValidHttpUrl(file)) {
		return file;
	}

	let absolutePath;

	if (path.isAbsolute(file)) {
		absolutePath = file;
	} else {
		absolutePath = path.resolve(dirname, file);
	}

	if (!fs.existsSync(absolutePath)) {
		throw new Error(`${absolutePath} not found`);
	}
	return absolutePath;
};

export function isValidHttpUrl(string: string) {
	let url;

	try {
		url = new URL(string);
	} catch {
		return false;
	}

	return url.protocol === 'http:' || url.protocol === 'https:';
}

export * from './types';
