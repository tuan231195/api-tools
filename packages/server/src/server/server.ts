import {
	Config,
	defaultValidateOptions,
	InvalidRequestError,
	InvalidResponseError,
	ValidateOptions,
} from 'src/server/types';
import morgan from 'morgan';
import bodyParser from 'body-parser';
import errorHandler from 'errorhandler';
import {
	OpenAPI,
	parse,
	ParsedOperationSchemaWithInfo,
	ramlToExpress,
} from '@vdtn359/api-tools-core';
import fs from 'fs';
import requireDir from 'require-dir';
import { fakeOperationResponse, initSeed } from '@vdtn359/api-tools-faker';
import { camelCase, mapKeys } from 'lodash';
import { Express, NextFunction, Request, Response } from 'express';
import logger from 'src/logger';
import { ValidationErrorItem } from 'express-openapi-validator/dist/framework/types';

export const setupServer = async (config: Config) => {
	logger('Running with config', config);
	if (config.seed != null) {
		initSeed(config.seed);
	}

	const express = require('express');
	const app = express();
	const router = express.Router();
	app.set('port', config.port || 3000);
	app.use(morgan('combined', { stream: { write: (msg) => logger(msg) } }));
	app.use(bodyParser.json());
	app.use(bodyParser.urlencoded({ extended: true }));

	let cleanupFunction: Function | null = null;
	if (config.init && fs.existsSync(config.init)) {
		const init = require(config.init).default;
		logger('Call custom init');
		cleanupFunction = await init(app);
	}

	const { schema: parsedSchema, document } = await parse(config.uri);
	if (config.validate) {
		let validateOptions: ValidateOptions;
		if (typeof config.validate === 'boolean') {
			validateOptions = defaultValidateOptions;
		} else {
			validateOptions = {
				...defaultValidateOptions,
				...config.validate,
			};
		}
		logger('Init swagger validator middleware');
		initSwaggerMiddleware({ app, document, validateOptions, config });
	}

	if (config.routes && fs.existsSync(config.routes)) {
		const setupExtraRoutes = require(config.routes).default;
		logger('Call custom setup routes');
		await setupExtraRoutes(router);
	}
	let overrides: Record<string, any> = {};
	if (config.overrides && fs.existsSync(config.overrides)) {
		overrides = mapKeys(
			requireDir(config.overrides, {
				recurse: false,
				extensions: ['.js', '.ts'],
			}),
			(value, key) => camelCase(key)
		);
	}

	const apiRouter = express.Router();
	await Promise.all(
		Object.entries(parsedSchema).map(async ([operationId, operation]) => {
			logger(`Setup ${operationId}`);
			const expressUrl = ramlToExpress(operation.info.path);
			const overrideFunction =
				overrides[operationId] &&
				typeof overrides[operationId].default === 'function'
					? overrides[operationId].default
					: undefined;

			apiRouter[operation.info.method](
				expressUrl,
				getRouteHandler(operation, config.seed, overrideFunction)
			);
		})
	);
	router.use(config.baseUrl, apiRouter);
	app.use(router);

	if (cleanupFunction) {
		cleanupFunction();
	}
	app.use(errorHandler());

	return app;
};

const getRouteHandler = (
	operation: ParsedOperationSchemaWithInfo,
	seed?: string,
	overrideFunc?: Function
) => {
	const defaultStatusStr = Object.keys(
		operation.schema.responseBody.statuses
	).find((key) => {
		const statusCode = parseInt(key, 10);
		return statusCode >= 200 && statusCode < 300;
	});
	const responseStatus = defaultStatusStr
		? parseInt(defaultStatusStr, 10)
		: 200;

	const defaultResponseGetter = (() => {
		let cachedDefaultResponse: any;

		return {
			async get(): Promise<any> {
				if (cachedDefaultResponse) {
					return cachedDefaultResponse;
				}
				const defaultResponse = defaultStatusStr
					? await fakeOperationResponse({
							schema: operation.schema,
							status: responseStatus,
					  })
					: undefined;
				if (seed != null) {
					cachedDefaultResponse = defaultResponse;
					return cachedDefaultResponse;
				} else {
					return defaultResponse;
				}
			},
		};
	})();

	return async (request: Request, response: Response, next: NextFunction) => {
		let defaultResponse: any;
		try {
			defaultResponse = await defaultResponseGetter.get();
		} catch (e) {
			defaultResponse = null;
		}

		try {
			let finalResponse = defaultResponse;
			if (overrideFunc) {
				logger(
					`Use override handler for ${operation.info.operationId}`
				);
				finalResponse = await overrideFunc({
					request,
					response,
					fakeResponse: defaultResponse,
				});
			}
			if (!response.headersSent) {
				if (finalResponse != undefined) {
					if (finalResponse && typeof finalResponse === 'object') {
						response.status(responseStatus).json(finalResponse);
					} else {
						response.status(responseStatus).send(finalResponse);
					}
				} else {
					response.sendStatus(responseStatus);
				}
			}
		} catch (err) {
			logger(`Route handler ${operation.info.operationId} failed`, err);
			next(err);
		}
	};
};

function initSwaggerMiddleware({
	app,
	document,
	validateOptions,
	config,
}: {
	app: Express;
	document: OpenAPI.Document;
	validateOptions: ValidateOptions;
	config: Config;
}) {
	if ('swagger' in document) {
		const validator = require('swagger-express-validator');
		const schema = {
			...document,
			basePath: config.baseUrl,
		};
		delete schema.definitions;
		// swagger v2
		app.use(
			validator({
				...validateOptions,
				returnRequestErrors: validateOptions.validateRequests,
				returnResponseErrors: validateOptions.validateResponses,
				schema,
				requestValidationFn: (
					req: Request,
					data: any,
					errors: any[]
				) => {
					throw new InvalidRequestError(
						ajvErrorsToValidatorError(errors)
					);
				},
				responseValidationFn: (
					req: Request,
					data: any,
					errors: any[]
				) => {
					throw new InvalidResponseError(
						ajvErrorsToValidatorError(errors)
					);
				},
			})
		);
	} else {
		// swagger v3
		const OpenApiValidator = require('express-openapi-validator');
		app.use(
			OpenApiValidator.middleware({
				...validateOptions,
				apiSpec: {
					...document,
					servers: [
						{
							url: config.baseUrl,
						},
					],
				},
			})
		);

		app.use(
			(err: Error, req: Request, res: Response, next: NextFunction) => {
				if (err instanceof OpenApiValidator.error.BadRequest) {
					return next(new InvalidRequestError((err as any).errors));
				}
				if (err instanceof OpenApiValidator.error.InternalServerError) {
					return next(new InvalidResponseError((err as any).errors));
				}
				next(err);
			}
		);
	}
}

function ajvErrorsToValidatorError(errors: any[]): ValidationErrorItem[] {
	return errors.map((e) => {
		const params: any = e.params;
		const required =
			params?.missingProperty &&
			e.dataPath + '.' + params.missingProperty;
		const additionalProperty =
			params?.additionalProperty &&
			e.dataPath + '.' + params.additionalProperty;
		const path =
			required ?? additionalProperty ?? e.dataPath ?? e.schemaPath;
		return {
			path,
			message: e.message,
			error_code: `${e.keyword}.openapi.validation`,
		};
	});
}
