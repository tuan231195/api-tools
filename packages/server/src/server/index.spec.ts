import { Express } from 'express';
import path from 'path';
import Chance from 'chance';
import { initServer } from 'src/server';
import request from 'supertest';

const chance = new Chance();

describe('initServer', () => {
	let app: Express;

	describe.each(['config.v2.json', 'config.v3.json'])('%s', (file) => {
		beforeAll(async () => {
			app = await initServer(
				path.resolve(__dirname, '..', '..', 'example', file),
				'123'
			);
		});

		it('should allow custom routes', async () => {
			const response = await request(app).get('/hello').expect(200);
			expect(response.text).toEqual('ok');
		});

		it('should allow custom middlewares', async () => {
			const response = await request(app).get('/hello').expect(200);
			expect(response.headers['x-powered-by']).toEqual(
				'mock-swagger-server'
			);
		});

		it('should perform request validation', async () => {
			const response = await request(app)
				.post('/api/user')
				.send({})
				.expect(400);
			expect(response.body).toMatchSnapshot();
		});

		it('should allow valid request and return mock response', async () => {
			const response = await request(app)
				.post('/api/user')
				.send({
					id: chance.guid(),
					firstName: chance.first(),
					lastName: chance.last(),
					username: chance.word(),
					password: chance.word(),
				})
				.expect(200);
			expect(response.body).toMatchSnapshot();
		});

		it('should allow customised response', async () => {
			const response = await request(app)
				.get('/api/pet/findByStatus')
				.query({
					status: 'available',
				})
				.expect(200);
			expect(response.body).toMatchSnapshot();
		});
	});
});
