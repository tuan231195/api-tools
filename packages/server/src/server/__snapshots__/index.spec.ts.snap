// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initServer config.v2.json should allow customised response 1`] = `
Array [
  Object {
    "category": Object {
      "id": "502063d8-dee7-56f0-b3db-23265defdd72",
      "name": "Walter Simon",
    },
    "id": "d6211c27-716a-50d2-bd43-a69d9a3b9993",
    "name": "Pet 0 (Query available)",
    "photoUrls": Array [
      "http://iDAlgGHlXZTjgoIFlrrvwrLtkaWEayktr.hqeVNDhHTOPNiD-8,A.",
      "http://CEzGTIVnwzRu.oqdqg",
      "http://XqOorATjnNtIrsxOYOaDdDolIRtxEzAf.neloDlNOwfOXnRuiyvwkZYU",
      "https://qymEgnnquJXjvzoHQgxONi.wirneD+qGTf8frTZBeKoPpYDmW9I3H7mWuRk7wOC3,0I+F6rXEVX-3MxAwT1A0HCWwe20N1oxIzHWV3nowlUt0z",
    ],
    "status": "sold",
    "tags": Array [
      Object {
        "id": "7b34fc0c-45d0-59e0-a099-a2c3205267aa",
        "name": "Isaac McDonald",
      },
      Object {
        "id": "16c59251-6457-55ef-96e2-2af5e2a55b39",
        "name": "Steve Ortega",
      },
      Object {
        "id": "72aec4e8-4bac-52cc-818e-02001a33ca75",
        "name": "Katharine Kennedy",
      },
      Object {
        "id": "43b526a4-ddd6-5ae2-9477-6b43e0ac3e00",
        "name": "Florence Wise",
      },
      Object {
        "id": "81b89e2c-e652-510b-b507-913127275db6",
        "name": "Loretta Dean",
      },
    ],
  },
  Object {
    "category": Object {
      "id": "584e99be-3faf-5b33-9337-15e51ab1894f",
      "name": "Norman Allen",
    },
    "id": "3843bf31-d60c-5215-8aa1-5c1cdb1c6152",
    "name": "Pet 1 (Query available)",
    "photoUrls": Array [
      "http://Plt.lduwckBB8YSAhvlavdPtnENKuNsVMRvgFk.qyxFDe1+V4RMG",
    ],
    "status": "sold",
    "tags": Array [
      Object {
        "id": "af415e5a-c7cb-5660-b4b0-71c98c0ec208",
        "name": "Josie Rice",
      },
    ],
  },
  Object {
    "category": Object {
      "id": "c925c626-3ace-59b3-99cd-3a59fb11d2da",
      "name": "Josie Barber",
    },
    "id": "74d4049f-39a3-59d8-9256-f8cd36c98ec5",
    "name": "Pet 2 (Query available)",
    "photoUrls": Array [
      "https://fswiLEUgeBKDJWjfYbKtZJApZOTJgt.lsieLAaUFLZ.ZNyK-Sz,8lXQ7ccZOXrR,qA3Qe+W2UnnR1y-2kIbypWwJftDcPbr8xEA",
      "https://UVPYcbPH.fdxhx4JY3qxpiF0Q5,vxpEYUoX-",
      "http://aiENwRTqshYQL.akghbNhBc",
      "https://qiZYWhWmcBCvitFvFQnsSzXnfRHY.ujt3JTDP,R",
    ],
    "status": "sold",
    "tags": Array [
      Object {
        "id": "0f73ffc8-a597-5d6c-9d2e-09612127442a",
        "name": "Eunice Gutierrez",
      },
      Object {
        "id": "58f07066-bd93-5321-9a8f-c4afed5315cf",
        "name": "Dorothy Mendez",
      },
      Object {
        "id": "173baf8a-d4cb-54fc-9bb3-9792426ad885",
        "name": "Mike Summers",
      },
      Object {
        "id": "a751e0ae-4cc8-5651-83fd-5b612c5aee3a",
        "name": "Owen Lowe",
      },
    ],
  },
]
`;

exports[`initServer config.v2.json should allow valid request and return mock response 1`] = `Object {}`;

exports[`initServer config.v2.json should perform request validation 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'id'",
      "path": ".id",
    },
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'username'",
      "path": ".username",
    },
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'firstName'",
      "path": ".firstName",
    },
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'lastName'",
      "path": ".lastName",
    },
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'password'",
      "path": ".password",
    },
  ],
}
`;

exports[`initServer config.v3.json should allow customised response 1`] = `
Array [
  Object {
    "category": Object {
      "id": "a2a637dc-6055-5b6b-8783-b814be3ff241",
      "name": "Lena Schwartz",
    },
    "id": "402314d7-a4b6-54ca-a564-ac4c7f4b4c73",
    "name": "Pet 0 (Query available)",
    "photoUrls": Array [
      "https://jawv.huirPVG0sa",
      "http://OnPBPdViEFQiampArlmArMtSd.ltfqEygBdOw6B64T1MBfPR790m9rbwo2HZirV.HyLemuihRMxvgeXQ3f6fEfiIN+uUn9rH+UbphRTjA52y",
    ],
    "status": "available",
    "tags": Array [
      Object {
        "id": "ba73c070-96cd-5d38-a902-d3bf738537ac",
        "name": "Lettie Carter",
      },
      Object {
        "id": "fe2a84c0-1dc1-556d-a372-0d258b3508f9",
        "name": "Thomas Salazar",
      },
    ],
  },
  Object {
    "category": Object {
      "id": "c4e84bac-2cc3-518e-9200-1a33ca7555e4",
      "name": "Hattie Keller",
    },
    "id": "92516457-5ef9-56e2-baf5-e2a55b3972ae",
    "name": "Pet 1 (Query available)",
    "photoUrls": Array [
      "http://seGNFkXNslFlBZRInlIcTDAcl.brjTgFn0reV07Nhhobep-JZcTEP0t-5VL52qvcfYerVlfov8rSy5nu+K",
      "https://pOxAYVDwZbWOwNwPxUXkXRuyXi.vqhO59VCpybh0RxAZoD7RU3k2nfYiI1HpQhPticx3HC,4opiZn",
    ],
    "status": "available",
    "tags": Array [
      Object {
        "id": "9fc1199b-eaf3-5a74-88d7-6e033a9a0a8b",
        "name": "Travis McDaniel",
      },
      Object {
        "id": "50b30172-59af-5bdb-894e-5b49db32fe34",
        "name": "Jonathan Moore",
      },
      Object {
        "id": "7aec3eac-dd46-542c-a10c-37a1abb07b34",
        "name": "Eric Bryan",
      },
      Object {
        "id": "d09e0a09-9a2c-5320-8267-aaa4dc3716c5",
        "name": "Dustin Salazar",
      },
    ],
  },
  Object {
    "category": Object {
      "id": "8dee76f0-23db-5232-85de-fdd72e593843",
      "name": "Bettie Ford",
    },
    "id": "4ddd6ae2-9477-56b4-8e0a-c3e007a881b8",
    "name": "Pet 2 (Query available)",
    "photoUrls": Array [
      "https://lgGHlXZTjgoIFlrrv.isjGoK4unAfLv8CNxHTOPNiD-8,A.YL",
    ],
    "status": "pending",
    "tags": Array [
      Object {
        "id": "ce65210b-f507-5913-b272-75db6002d621",
        "name": "Maud Burke",
      },
      Object {
        "id": "1c27716a-0d2f-5d43-a69d-9a3b9993ad85",
        "name": "Estella Clarke",
      },
    ],
  },
  Object {
    "category": Object {
      "id": "ac7cb660-e4b0-571c-98c0-ec208abd74d4",
      "name": "Katie Guzman",
    },
    "id": "1d60c215-6aa1-55c1-adb1-c615223a584e",
    "name": "Pet 3 (Query available)",
    "photoUrls": Array [
      "http://erxXmGvKeVdrVKOvdDlGmOxmVMWBSAUjL.rfuKQiQGysnDlNOwfOXnRuiyvwkZYUYvZswH5mrXyIxDZYaLeLsoR3tMGHhwBDI",
      "https://qJAJRumDGVGWRfYCXGRwMEECjMvjPK.gzjzrC3,0R.X6rXEVN3WgxITwT1APC+SweRN1oD9IzHM3nowl7upzQuEpMiqnAT+jGrQ1gZXl4",
      "http://cGpnTyECpFo.prkcxPMZLgfrPiJcq8Y4ouBCa",
    ],
    "status": "pending",
    "tags": Array [
      Object {
        "id": "99be3faf-b337-5337-b5e5-1ab1894f72ea",
        "name": "Ricky Peters",
      },
    ],
  },
  Object {
    "category": Object {
      "id": "ffc8a597-d6c9-5d2e-9961-2127442a7345",
      "name": "Addie Allison",
    },
    "id": "f39a39d8-8256-5f8c-a36c-98ec56610f73",
    "name": "Pet 4 (Query available)",
    "photoUrls": Array [
      "http://pxb.amgxZT2He74dCZhBEzGZe8CoOXV2AaUFLOZNFqK-45z,VXHcc.LOXrR,nj32OeX2UnqSR1yYkIc8ypWwJ",
      "https://HlxYSB.vgjgx7hxCiM7hfIULJlg-bUy.THtZ5TiyVEUv",
    ],
    "status": "available",
    "tags": Array [
      Object {
        "id": "6bd93321-0a8f-5c4a-bed5-315cf173baf8",
        "name": "Mabelle Cannon",
      },
      Object {
        "id": "ad4cb4fc-9bb3-5979-b426-ad885f40a751",
        "name": "Jorge Harris",
      },
      Object {
        "id": "e4cc8651-53fd-55b6-b2c5-aee3ac93c925",
        "name": "Don Burton",
      },
      Object {
        "id": "63ace9b3-79cd-53a5-9fb1-1d2da9254119",
        "name": "Verna Christensen",
      },
      Object {
        "id": "7ee7372e-2bf4-582f-af2d-c5d51154d687",
        "name": "Mitchell Casey",
      },
    ],
  },
]
`;

exports[`initServer config.v3.json should allow valid request and return mock response 1`] = `
Object {
  "email": "dasinih@nadedok.ch",
  "firstName": "Jerry",
  "id": "8f5ff6a9-ad24-5798-bd11-f95c5eeb758e",
  "lastName": "Simoni",
  "password": "RNVhUantPjOu",
  "phone": "(526) 893-4707",
  "userStatus": true,
  "username": "ZBwiHHrAoRhZiENxHDXh",
}
`;

exports[`initServer config.v3.json should perform request validation 1`] = `
Object {
  "errors": Array [
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'id'",
      "path": ".body.id",
    },
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'username'",
      "path": ".body.username",
    },
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'firstName'",
      "path": ".body.firstName",
    },
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'lastName'",
      "path": ".body.lastName",
    },
    Object {
      "code": "InvalidRequestError",
      "message": "should have required property 'password'",
      "path": ".body.password",
    },
  ],
}
`;
